spec-tests/src/test/java/org/candlepin/spec/products/ProductVersioningSpecTest.java

```diff
@@ -0,0 +1,664 @@
+        assertEquals(prod6.getUuid(), prod5.getUuid());
+    }
+
+    @Test
+    public void shouldCleanupOrphansWithoutInterferingWithNormalActions() throws Exception {
+        // NOTE:
+        // This test takes advantage of the immutable nature of products with the in-place update branch
+        // disabled. If in-place updates are ever reenabled, we'll need a way to generate large numbers
+        // of orphaned products for this test.
+
+        OwnerDTO owner1 = ownerApi.createOwner(Owners.random());
+        OwnerDTO owner2 = ownerApi.createOwner(Owners.random());
+
+        String prefix = "test-product-";
+        int offset = 10000;
+        int length = 100;
+
+        // Repeat this test a few(ish) times to hopefully catch any synchronization error
+        for (int i = 0; i < 10; i++) {
+            final int base = offset;
+            List<String> o1Uuids = new ArrayList<>();
+            List<String> o2Uuids = new ArrayList<>();
+
+            // Create a bunch of dummy products
+            for (int j = base; j < base + length - 1; j++) {
+                String id = String.format("%s%d", prefix, j);
+                // create product and immediately update it to generate an orphaned product
+                ownerProductApi.createProductByOwner(owner1.getKey(), new ProductDTO().name(id).id(id));
+            }
+
+            // Attempt to update and create new products to get into some funky race conditions with
+            // convergence and orphanization
+            Thread updater = new Thread(() -> {
+                for (int j = base; j < base + length - 1; j++) {
+                    String id = String.format("%s%d", prefix, j);
+                    // create product and immediately update it to generate an orphaned product
+                    ProductDTO prod = ownerProductApi.updateProductByOwner(owner1.getKey(), id,
+                        new ProductDTO().name(id + "-update"));
+                    o1Uuids.add(prod.getUuid());
+                }
+            });
+            updater.start();
+            Thread generator = new Thread(() -> {
+                for (int j = base; j < base + length - 1; j++) {
+                    String id = String.format("%s%d", prefix, j);
+                    ProductDTO prod = ownerProductApi.createProductByOwner(owner2.getKey(),
+                        new ProductDTO().name(id).id(id));
+                    o2Uuids.add(prod.getUuid());
+                }
+            });
+            generator.start();
+
+            sleep(1000);
+            AsyncJobStatusDTO job = client.jobs().scheduleJob("OrphanCleanupJob");
+            updater.join();
+            generator.join();
+            AsyncJobStatusDTO jobStatus = client.jobs().waitForJob(job);
+            assertEquals("FINISHED", jobStatus.getState());
+
+            // Verify the products created/updated still exist
+            assertThat(o1Uuids.stream()
+                .filter(x -> client.products().getProduct(x) != null)
```

Comment:
If we're going to try to do this in one go like this, the assertion mapper would be a better option:

```java
assertThat(o1Uuids)
  .map(uuid -> client.products().getProduct(uuid))
  // other checks here
```

After the call to .map, the collection will contain product entities which can then be verified however necessary.